### PR TITLE
Fermo flask sh

### PR DESCRIPTION
**Pull request by shabicht**
- Implement file validity checking based on the filename and parse a session-file for basic info to be displayed in an overview table
- Display an "empty" table on the loading page prior to file upload
- Test the functions for filename validity check and saving

### CODE REVIEW

File: src/fermo/app_utils/input_testing.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    current_app,  # allows to access the config elements
+)
+from werkzeug.utils import secure_filename
+import os
+import json
+
+
+def check_file_format(filename, file_format, allowed_extensions):
+    ''' Check filename for validity
+
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:11]: Discuss doc-strings: do we want to include type hinting?

File: src/fermo/app_utils/input_testing.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    filename = secure_filename(file.filename)
+    if filename == '':
+        return 'No file was loaded. Please upload a session-file.', None
+    if file and not check_file_format(filename, file_format,
+                                      allowed_extensions):
+        try:
+            file.save(os.path.join(upload_folder, filename))
+        except FileNotFoundError as e:
+            print(e)
+            return "File or folder didn't exist, so the uploaded file couldn't\
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:65]: Better to not use abbreviations 😅

File: src/fermo/app_utils/input_testing.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    uploaded_version = content_dict['FERMO_version']
+    uploaded_version_split = uploaded_version.split('.')
+    major, minor, fix = (
+        uploaded_version_split[0],
+        uploaded_version_split[1],
+        uploaded_version_split[2],
+        )
+    if major != version.split('.')[0] or \
+       minor != version.split('.')[1]:
+        return f'''❌ Error: The loaded session file "{filename}"
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:95]: Better to use quotes to combine conditionals: 
```
if (
    major != version.split('.')[0] or minor != version.split('.')[1]
):
```

Comment (mmzdouc) [reply src/fermo/app_utils/input_testing.py:95]: (the current behaviour is not adherent to proper semantic versioning  - minor and fix should also be backward compatible - but lets keep it for now)

File: src/fermo/app_utils/input_testing.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    else:
+        return None
+
+
+def empty_loading_table():
+    '''Generate placeholder table for loading page
+
+    Returns
+    -------
+    list of lists: placeholder for table before file upload
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:115]: should be just `list`

File: src/fermo/app_utils/input_testing.py (hunk 1 above)
Comment (mmzdouc) [reply src/fermo/app_utils/input_testing.py:11]: Discuss Markdown

File: src/fermo/app_utils/input_testing.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    version: 'str'; current version of FERMO as read from __version__.py file
+
+    Return
+    ------
+    'dict': containing the parsed file info
+    'str': message for (possible) incompatibility or FileNotFoundError
+    '''
+    try:
+        with open(os.path.join(current_app.config.get('UPLOAD_FOLDER'),
+                               filename)) as f:
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:225]: style

Comment (mmzdouc) [reply src/fermo/app_utils/input_testing.py:225]: Also take a look at this handy package: https://docs.python.org/3/library/pathlib.html
Much better to work with than os

File: src/fermo/views.py
Comment (mmzdouc) [inline src/fermo/views.py:1]: MIght need some basic docstrings

File: src/fermo/views.py [hunk 6]
```diff
@@ -1,25 +1,21 @@
...
 )
-import __version__
-# import config
-from werkzeug.utils import secure_filename
-import os
+import fermo.__version__ as __version__
+from fermo.app_utils.input_testing import (
+    save_file,
+    parse_sessionfile,
+    empty_loading_table)
```
Comment (mmzdouc) [inline src/fermo/views.py:14]: style

File: src/fermo/views.py [hunk 7]
```diff
@@ -29,38 +25,45 @@ def landing(version=__version__.__version__):
...
-                    file.save(os.path.join('src/fermo/uploads/', filename))
-                except FileNotFoundError:
-                    print("file or folder didn't exist, so the uploaded file couldn't be saved")
+        if 'sessionFile' not in request.files:
+            print('Input ID was not in the request')
+        else:
+            sessionfile = request.files['sessionFile']
+            allowed_extensions = current_app.config.get('ALLOWED_EXTENSION')
+            upload_folder = current_app.config.get('UPLOAD_FOLDER')
+            message, filename = save_file(sessionfile, '.json',
```
Comment (mmzdouc) [inline src/fermo/views.py:34]: style

File: src/fermo/views.py [hunk 8]
```diff
@@ -29,38 +25,45 @@ def landing(version=__version__.__version__):
...
-                flash("File must be a json-file!")
+                flash(message)
                 return redirect(request.url)
-    return render_template('loading.html', version=version)
+    return render_template('loading.html', version=version,
+                           table=empty_loading_table())
 
 
-@views.route("/loading/<filename>")  # for inspecting the uploaded file before submitting it to the dashboard.
+@views.route("/loading/<filename>")  # for inspecting the uploaded file before
```
Comment (mmzdouc) [inline src/fermo/views.py:46]: docstring

File: src/fermo/views.py [hunk 9]
```diff
@@ -29,38 +25,45 @@ def landing(version=__version__.__version__):
...
+                           table=table_dict)
 
 
-@views.route("/dashboard")
-def dashboard(version=__version__.__version__):
-    return render_template('dashboard.html', version=__version__.__version__)
+@views.route("/processing", methods=['GET', 'POST'])
+def processing(version=__version__.__version__):
+    if request.method == 'POST':
+        filename = save_file(['peaktableFile', 'MSMSFile',
```
Comment (mmzdouc) [inline src/fermo/views.py:59]: style

File: tests/test_file_upload.py
Comment (mmzdouc) [inline tests/test_file_upload.py:1]: discuss style

File: tests/test_file_upload.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                         [('metadata.csv', '.csv'),
+                          ('sessionfile.json', '.json'),
+                          ('MSMS.mgf', '.mgf')
+                          ])
+def test_check_file_format_valid(filename, file_format, allowed_extensions):
+    '''test valid input for check_file_format()'''
+    assert not check_file_format(filename, file_format, allowed_extensions)
+
+
+@pytest.mark.parametrize('filename, file_format',
```
Comment (mmzdouc) [inline tests/test_file_upload.py:61]: can be reduced to a single example

File: src/fermo/app_utils/input_testing.py (hunk 1 above)
Comment (mmzdouc) [reply src/fermo/app_utils/input_testing.py:11]: Discuss implicit/explicit returns

File: tests/test_file_upload.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    assert check_file_format(filename, file_format, allowed_extensions)
+
+
+@pytest.mark.parametrize('filename, file_format',
+                         [('metadata.txt', '.txt'),
+                          ('sessionfile.json', 'json')
+                          ])
+def test_check_file_format_raise(filename, file_format, allowed_extensions):
+    '''test input that should raise an error in check_file_format()'''
+    with pytest.raises(ValueError):
```
Comment (mmzdouc) [inline tests/test_file_upload.py:79]: Discuss

File: tests/test_file_upload.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        check_file_format(filename, file_format, allowed_extensions)
+
+
+def test_save_file_successfully(allowed_extensions, session_file,
+                                upload_folder):
+    _, filename = save_file(session_file, '.json', allowed_extensions,
```
Comment (mmzdouc) [inline tests/test_file_upload.py:85]: Underscore?

File: src/fermo/app_utils/input_testing.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+        File format of one specific input with the leading dot
+    allowed_extension: `list`
+        List of strings as specified in config file
+
+    Return
+    ------
+    feedback: `str`
+        String with feedback for the user to be flashed
+    '''
+    if '.' not in file_format:
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:30]: unit tests must not be in production code - please remove and create separate test cases

File: src/fermo/app_utils/input_testing.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+            file_format,
+            allowed_extensions,
+        )
+    if feedback:
+        return feedback, False
+    else:
+        try:
+            file.save(os.path.join(upload_folder, filename))
+            return filename, True
+        except FileNotFoundError as e:
```
Comment (mmzdouc) [inline src/fermo/app_utils/input_testing.py:92]: Saving into folder is only a temporary solution, right?

File: tests/test_file_upload.py
Comment (mmzdouc) [inline tests/test_file_upload.py:1]: Temporary functions (e.g. saving to folder) do require testing. Chances are that they stay behind in production and need to be weeded out afterwards.

File: src/fermo/app_utils/input_testing.py (hunk 14 above)
Comment (shabicht) [reply src/fermo/app_utils/input_testing.py:92]: yes, the upload folder as set in the configurations is supposed to be replaced by a database. I'm not yet sure how the connection to a database will be implemented therefore this temporary solution

File: src/fermo/app_utils/input_testing.py (hunk 13 above)
Comment (shabicht) [reply src/fermo/app_utils/input_testing.py]: upon further inspection this code turned out to be unnecessary and was removed

**Pull request merged**